llama: refine developer docs (#7121)
ollama/ollama: docs/development.md

# Development

> [!IMPORTANT]

> > [Transitional new Go llama Runner](#transitional-new-go-llama-runner)

Reviewer: ```suggestion
> The `llm` package that loads and runs models is being updated to use a new [Go runner](#transition-to-go-runner): this should only impact a small set of PRs however it does change how the project is built.
```

```

You will need to ensure your PATH includes go, cmake, gcc and clang mingw32-make to build ollama from source. (typically `C:\msys64\clangarm64\bin\`)

> ## Transitional new Go llama Runner

Reviewer: ```suggestion
## Transition to Go runner
```

> The Ollama team is working on moving to a new Go based llama runner subprocess.  During a transition period, this new Go runner is "opt in" at build time, and requires using a different approach to build.

Reviewer: ```suggestion
The Ollama team is working on moving to a new Go based runner that loads and runs models in a subprocess to replace the previous code under `ext_server`. During this transition period, this new Go runner is "opt in" at build time, and requires using a different approach to build:
```

> To build the C++ based runner use:

Reviewer: Would swap these and:

```suggestion
To build the existing C++ based runner use:
```

```
go generate ./...
go build .
```

To build the new Go runner, use: (Adjust the job count based on your number of processors for a faster build)

```
make -C llama -j 5
go build .
```

> After we complete the transition to use the Go server exclusively, both `make` and `go generate` will build the Go Runner.  The instructions in this document assume an "opt in" build of the new Go server.

Reviewer: ```suggestion
After the transition to use the Go server exclusively, both `make` and `go generate` will build the Go runner.
```
